File: 2025/1/2/golang/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

type Dial struct{ value, zeros int }

func NewDial(v int) *Dial { return &Dial{value: v % 100} }

func Abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func (d *Dial) Left(n int) {
	// fmt.Println("Turning left:", n)
	if d.value-n <= 0 {
		add := (n - d.value) / 100
		if d.value != 0 {
			add++
		}
		d.zeros += add
		// if add > 0 {
		//	fmt.Println("Crossed zero", add, "times, bringing the zero count to", d.zeros)
		// }
	}
	d.value = (d.value - n) % 100
	if d.value < 0 {
		d.value += 100
	}
}

func (d *Dial) Right(n int) {
	// fmt.Println("Turning right:", n)
	if d.value+n >= 100 {
		add := (n - (100 - d.value)) / 100
		if d.value != 0 || n >= 100 {
			add++
		}
		d.zeros += add
		// if add > 0 {
		// 	fmt.Println("Crossed zero", add, "times, bringing the zero count to", d.zeros)
		// }
	}
	d.value = (d.value + n) % 100
}

func main() {
	d := NewDial(50)
	sc := bufio.NewScanner(os.Stdin)

	for sc.Scan() {
		mov := sc.Text()
		if mov == "" {
			break
		}
		// fmt.Println("Read ", mov)
		dir := mov[0]
		n, _ := strconv.Atoi(mov[1:])
		switch dir {
		case 'L':
			d.Left(n)
		case 'R':
			d.Right(n)
		}
		// fmt.Println("The value is:", d.value)
	}

	fmt.Println("The password is:", d.zeros)
}

File: 2025/1/2/python/main.py
import sys


class Dial:
    def __init__(self):
        self.value = 50
        self.zeros = 0

    def left(self, val):
        print(f"Turning left: {val}")
        if self.value - val <= 0:
            self.zeros += int(self.value != 0) + ((val - self.value) // 100)
            if int(self.value != 0) + ((val - self.value) // 100) > 0:
                print(
                    f"Crossed zero {int(self.value != 0) + ((val - self.value) // 100)} times, bringing the zero count to {self.zeros}"
                )
        self.value = (self.value - val) % 100

    def right(self, val):
        print(f"Turning right: {val}")
        if self.value + val >= 100:
            add_zeros = int(self.value != 0 or val >= 100) + (
                (val - (100 - self.value)) // 100
            )
            self.zeros += add_zeros
            if add_zeros > 0:
                print(
                    f"Crossed zero {add_zeros} times, bringing the zero count to {self.zeros}"
                )
        self.value = (self.value + val) % 100


if __name__ == "__main__":
    d = Dial()
    for mov in sys.stdin.readlines():
        if mov.strip() == "":
            break
        if mov[0] == "L":
            d.left(int(mov[1:]))
        else:
            d.right(int(mov[1:]))
        print(f"The value is: {d.value}")
    print(f"The password is {d.zeros}")
